web/static/js/explore.js

```diff
@@ -2,255 +2,254 @@
-                    }
+$.fn.dataTable.pipeline = function (opts) {
+    // Configuration options
+    const conf = $.extend({
+        contentType: null,
+        pages: 5,     // number of pages to cache
+        url: '',      // script url
+        data: null,   // function or object with parameters to send to the server
+                      // matching how `ajax.data` works in DataTables
+        dataSrc: null,
+        method: 'GET' // Ajax HTTP method
+    }, opts);
+    // Private variables for storing the cache
+    let cacheLower = -1;
+    let cacheUpper = null;
+    let cacheLastRequest = null;
+    let cacheLastJson = null;
+
+    return function (request, drawCallback, settings) {
+        let ajax = false;
+        let requestStart = request.start;
+        const drawStart = request.start;
+        const requestLength = request.length;
+        const requestEnd = requestStart + requestLength;
+        if (settings.clearCache) {
+            // API requested that the cache be cleared
+            ajax = true;
+            settings.clearCache = false;
+        } else if (cacheLower < 0 || requestStart < cacheLower || requestEnd > cacheUpper) {
+            // outside cached data - need to make a request
+            ajax = true;
+        } else if (JSON.stringify(request.order) !== JSON.stringify(cacheLastRequest.order) ||
+            JSON.stringify(request.columns) !== JSON.stringify(cacheLastRequest.columns) ||
+            JSON.stringify(request.search) !== JSON.stringify(cacheLastRequest.search)
+        ) {
+            // properties changed (ordering, columns, searching)
+            ajax = true;
+        }
+
+        // Store the request for checking next time around
+        cacheLastRequest = $.extend(true, {}, request);
+
+        let json;
+        if (ajax) {
+            // Need data from the server
+            if (requestStart < cacheLower) {
+                requestStart = requestStart - (requestLength * (conf.pages - 1));
+
+                if (requestStart < 0) {
+                    requestStart = 0;
+                }
+            }
+
+            cacheLower = requestStart;
+            cacheUpper = requestStart + (requestLength * conf.pages);
+
+            request.start = requestStart;
+            request.length = requestLength * conf.pages;
+            conf.data.skip = requestStart;
+            conf.data.limit = request.length;
+            if ($("#datatable_filter > label > input")[0].value) {
+                conf.data.filter = $("#datatable_filter > label > input")[0].value;
+            }
+
+            // Provide the same `data` options as DataTables.
+            if (typeof conf.data === 'function') {
+                // As a function it is executed with the data object as an arg
+                // for manipulation. If an object is returned, it is used as the
+                // data object to submit
+                const d = conf.data(request);
```

Comment:
shouldn't the variable name be descriptive other than "d"? something like data or manipulation_data